### PR TITLE
Sets title of terminal to 'Helix Editor'

### PR DESCRIPTION
**Pull request by sonicrules1234**
Fixes https://github.com/helix-editor/helix/issues/2436

### CODE REVIEW

**Comment by the-mikedavis** [comment]
I would prefer that this reads as `hx <file>` if possible. You could use https://github.com/helix-editor/helix/commit/a1228e0aa1757e5c15da18be4ff2920c93454eec as a reference

**Comment by sonicrules1234** [comment]
Is it okay if I add a function that takes the Document object and sets the terminal accordingly to helix-view/src/lib.rs?

**Comment by kirawi** [comment]
You can use `Document::path()`

**Comment by dead10ck** [comment]
> I would prefer that this reads as `hx <file>` if possible. You could use [a1228e0](https://github.com/helix-editor/helix/commit/a1228e0aa1757e5c15da18be4ff2920c93454eec) as a reference

I personally think just "hx" is preferable. The problem with "hx <file>" is that you might have a long editing session where you open tons of other files, and/or close the first one you started with, so the title becomes misleading and confusing.

Even better might be to programmatically set it to positional argument 0, so that it comes out to the name of the command. On most systems, this would be "hx", but if someone prefers a different name for the executable, or their distro renames it.

**Comment by EricCrosson** [comment]
> I personally think just "hx" is preferable. The problem with `hx <file>` is that you might have a long editing session where you open tons of other files, and/or close the first one you started with, so the title becomes misleading and confusing.

Can we display `hx <name-of-active-buffer>`?

**Comment by dead10ck** [comment]
> Can we display `hx <name-of-active-buffer>`? Which would be the filename if the view is showing a file

That would make sense, though that would require a different approach, closer to the branch that was linked to earlier.

File: helix-term/src/application.rs [hunk 1]
```diff
@@ -167,9 +169,11 @@ impl Application {
                 // does not affect views without pos since it is at the top
                 let (view, doc) = current!(editor);
                 align_view(doc, view, Align::Center);
+                crate::set_title_from_doc(doc);
             }
         } else if stdin().is_tty() {
             editor.new_file(Action::VerticalSplit);
+            execute!(std::io::stdout(), terminal::SetTitle("hx [scratch]")).unwrap();
```
[the remark this answers is not in the document]
Comment (sonicrules1234) [reply helix-term/src/application.rs:176]: I'd like to get it working before we worry about that

File: helix-term/src/lib.rs [hunk 2]
```diff
@@ -22,3 +22,18 @@ fn true_color() -> bool {
...
+            "settitle hx {}",
+            doc.relative_path()
+                .as_deref()
+                .unwrap_or(std::path::Path::new("[scratch]"))
+                .to_str()
+                .unwrap() //,
+        ))
+    )
+    .unwrap();
+}
```
Comment (pickfire) [inline helix-term/src/lib.rs:39]: Missing eof

File: helix-term/src/lib.rs [hunk 3]
```diff
@@ -22,3 +22,18 @@ fn true_color() -> bool {
...
+pub fn set_title_from_doc(doc: &helix_view::Document) {
+    crossterm::execute!(
+        std::io::stdout(),
+        crossterm::terminal::SetTitle(format!(
+            "settitle hx {}",
+            doc.relative_path()
+                .as_deref()
+                .unwrap_or(std::path::Path::new("[scratch]"))
+                .to_str()
+                .unwrap() //,
```
Comment (pickfire) [inline helix-term/src/lib.rs:35]: What does this comment means?

File: helix-term/src/lib.rs [hunk 4]
```diff
@@ -22,3 +22,18 @@ fn true_color() -> bool {
 fn true_color() -> bool {
     true
 }
+
+pub fn set_title_from_doc(doc: &helix_view::Document) {
+    crossterm::execute!(
+        std::io::stdout(),
+        crossterm::terminal::SetTitle(format!(
+            "settitle hx {}",
```
Comment (pickfire) [inline helix-term/src/lib.rs:30]: Are you still testing with this?

**Comment by pickfire** [comment]
Converting this to work in progress since this looks like work in progress.

File: helix-term/src/lib.rs (hunk 4 above)
Comment (sonicrules1234) [reply helix-term/src/lib.rs:30]: Yes, because it works in some places, but not others

**Comment by JakeHL** [comment]
This feature might be worth waiting until hooks are implemented. If that's still a goal, anyway. Since not all terminals set their titles in the same way, or people might be using multiplexers, window managers etc. that behave differently.

If they can just attach to a buffer/file/directory changed hook they can set the title any way they choose.

**Comment by jfaz1** [comment]
> This feature might be worth waiting until hooks are implemented. If that's still a goal, anyway. Since not all terminals set their titles in the same way, or people might be using multiplexers, window managers etc. that behave differently.
> 
> If they can just attach to a buffer/file/directory changed hook they can set the title any way they choose.

Honestly I think `crossterm::terminal::SetTitle()` will work for the vast majority of users. If anything maybe a config option to turn it off and override with a hook, but this seems like basic enough functionality that should be included out-of-the-box, and it seems in line with Helix's goal. Out of curiosity on Windows I tried on `CMD`, `PowerShell`, `WezTerm`, and `Windows Terminal` and the window was renamed appropriately for every single one.

**Comment by pascalkuthe** [comment]
closing this one as stale. Thank you for contributing!

**Pull request closed without merging**